### PR TITLE
Use frame url for scripts without sources

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/29585

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Steps from issue

### CODE REVIEW

**Review by goodov — APPROVED** [review]

**Comment by pes10k** [comment]
```  if (secondary_url.SchemeIs(url::kDataScheme)) {
    secondary_url = render_frame()->GetWebFrame()->GetDocument().Url();
  }
```

What happens if the script is being executed in an iframe with URL of `about:blank`?  Will this report `about:blank` or the URL of the first non-`about:blank` parent frame?

**Comment by spylogsster** [comment]
> ```
>     secondary_url = render_frame()->GetWebFrame()->GetDocument().Url();
>   }
> ```
> 
> What happens if the script is being executed in an iframe with URL of `about:blank`? Will this report `about:blank` or the URL of the first non-`about:blank` parent frame?

should report `about:blank`

**Comment by pes10k** [comment]
I dont think thats correct then.  I will double check, but (I believe) a script in the example I gave should count as executing in the first non-about:blank parent frame

**Comment by pes10k** [comment]
yes, I think im correct, you want to base things off of whatever `document.domain` reports, which, in the case of a nested `about:blank` iframe, will be the domain of the first non-about:blank parent frame

File: components/content_settings/renderer/brave_content_settings_agent_impl.cc [hunk 1]
```diff
@@ -224,7 +224,15 @@ bool BraveContentSettingsAgentImpl::AllowStorageAccessSync(
 bool BraveContentSettingsAgentImpl::AllowScriptFromSource(
     bool enabled_per_settings,
     const blink::WebURL& script_url) {
-  const GURL secondary_url(script_url);
+  GURL secondary_url(script_url);
+  // For scripts w/o sources it should report the domain / site used for
+  // executing the frame (which most, but not all, of the time will just be from
+  // document.location
+  if (secondary_url.SchemeIs(url::kDataScheme)) {
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/content_settings/renderer/brave_content_settings_agent_impl.cc]: `SchemeIsLocal` looks good, thanks

**Comment by pes10k** [comment]
Are there tests for the following cases?  If so, then I think this is good:

1. when scripts are blocked for example.com, that scripts blocking in a local iframe embedded in example.com are also blocked
2. 1. when scripts are _not_ blocked for example.com, that scripts blocking in a local iframe embedded in example.com are _not_ blocked

**Comment by spylogsster** [comment]
> Are there tests for the following cases? If so, then I think this is good:
> 
> 1. when scripts are blocked for example.com, that scripts blocking in a local iframe embedded in example.com are also blocked
> 2. 1. when scripts are _not_ blocked for example.com, that scripts blocking in a local iframe embedded in example.com are _not_ blocked

yes, they are using a.com with a script without source and with `load_js_dataurls.js` script which creates about:blank iframe with additional script without source and checks they are blocked or not:
```
a.com
   <script>without source</script>
   load_js_dataurls.js
      <iframe src=about:blank>
          <script>without source</script>             
```

**Review by pes10k — APPROVED** [review]
Lgtm with the additional tests

File: components/content_settings/renderer/brave_content_settings_agent_impl.cc [hunk 2]
```diff
@@ -224,7 +224,15 @@ bool BraveContentSettingsAgentImpl::AllowStorageAccessSync(
 bool BraveContentSettingsAgentImpl::AllowScriptFromSource(
     bool enabled_per_settings,
     const blink::WebURL& script_url) {
-  const GURL secondary_url(script_url);
+  GURL secondary_url(script_url);
+  // For scripts w/o sources it should report the domain / site used for
+  // executing the frame (which most, but not all, of the time will just be from
+  // document.location
```
Comment (stephendonner) [inline components/content_settings/renderer/brave_content_settings_agent_impl.cc:230]: Uber-nit, here: you forgot the closing `)` in the source comment, after `// document.location)`

**Comment by stephendonner** [comment]
## Verified `PASSED` using 

Brave | 1.52.39 Chromium: 112.0.5615.49 (Official Build) nightly (64-bit)
-- | --
Revision | bd2a7bcb881c11e8cfe3078709382934e3916914-refs/branch-heads/5615@{#936}
OS | Windows 10 Version 22H2 (Build 19045.2846)

### Steps:
Followed those from https://github.com/brave/brave-browser/issues/29585#issue-1659431370

#### Confirmed the scripts which previously showed from `null` are now showing from `https://apps.facebook.com`

example | example | example | example | example
----------|-----------|----------|-----------|----------
image | image | image | image | image

**Comment by stephendonner** [comment]
@spylogsster mind getting an uplift going to `1.51.x`, for the above?  Thanks!

**Pull request merged**